server/src/main/java/org/elasticsearch/cluster/routing/RoutingNode.java

```diff
@@ -73,6 +78,7 @@ private RoutingNode(RoutingNode original) {
         this.initializingShards = new LinkedHashSet<>(original.initializingShards);
         this.startedShards = new LinkedHashSet<>(original.startedShards);
         this.shardsByIndex = Maps.copyOf(original.shardsByIndex, HashSet::new);
+        this.indexUuids = original.shardsByIndex.keySet().stream().map(Index::getUUID).collect(Collectors.toSet());
```

Comment:
Thinking out loud: I am worried this will have a noticeable performance impact. It looks like we construct this at least on every routing allocation (code path starts [here](https://github.com/elastic/elasticsearch/blob/e69a05bb9e57a83036ba791acc15b6678f807794/server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java#L763-L763)). I am worried that we have to loop over all the indices on that node again (we're also doing that on the line above). Perhaps we could combine this line with the one above and create one loop that fills both sets?

I _think_ the memory overhead of the `Set` is not that bad, as it's only the set itself that we hold in memory - the UUID strings themselves are already elsewhere on the heap. What do you think?

Reply:
That is a good point, what we can do is to run it in the invariant loop. This will make it slower for tests but it won't affect production code that runs without the assertions